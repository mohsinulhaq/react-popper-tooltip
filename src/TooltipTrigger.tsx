/**
 * @author Mohsin Ul Haq <mohsinulhaq01@gmail.com>
 */
import PopperJS from 'popper.js';
import React, {Component} from 'react';
import {createPortal} from 'react-dom';
import {Manager, Popper, Reference} from 'react-popper';
import Tooltip from './Tooltip';
import {
  GetTriggerPropsArg,
  TooltipTriggerProps,
  TooltipTriggerState,
  TriggerTypes
} from './types';
import {callAll, canUseDOM, noop} from './utils';

const DEFAULT_MODIFIERS: PopperJS.Modifiers = {
  preventOverflow: {
    boundariesElement: 'viewport'
  }
};

class TooltipTrigger extends Component<
  TooltipTriggerProps,
  TooltipTriggerState
> {
  public static defaultProps = {
    closeOnOutOfBoundaries: true,
    defaultTooltipShown: false,
    delayHide: 0,
    delayShow: 0,
    followCursor: false,
    onVisibilityChange: noop,
    placement: 'right',
    portalContainer: canUseDOM() ? document.body : null,
    trigger: 'hover',
    usePortal: canUseDOM()
  };

  public state: TooltipTriggerState = {
    tooltipShown: this.props.defaultTooltipShown
  };

  private hideTimeout?: number;
  private showTimeout?: number;
  private popperOffset?: PopperJS.Offset;

  public componentWillUnmount() {
    this.clearScheduled();
  }

  public render() {
    const {
      children,
      tooltip,
      placement,
      trigger,
      getTriggerRef,
      modifiers,
      closeOnOutOfBoundaries,
      usePortal,
      portalContainer,
      followCursor,
      getTooltipRef,
      ...restProps
    } = this.props;

    const popper = (
      <Popper
        innerRef={getTooltipRef}
        placement={placement}
        modifiers={{
          ...DEFAULT_MODIFIERS,
          ...(followCursor && {
            followCursorModifier: {
              enabled: true,
              fn: data => {
                this.popperOffset = data.offsets.popper;
                return data;
              },
              order: 1000
            }
          }),
          ...modifiers
        }}
        {...restProps}
      >
        {({
          ref,
          style,
          // tslint:disable-next-line
          placement,
          arrowProps,
          outOfBoundaries,
          scheduleUpdate
        }) => {
          if (followCursor && this.popperOffset) {
            const {pageX, pageY} = this.state;
            const {width, height} = this.popperOffset;
            const x =
              pageX! + width > window.scrollX + document.body.offsetWidth
                ? pageX! - width
                : pageX;
            const y =
              pageY! + height > window.scrollY + document.body.offsetHeight
                ? pageY! - height
                : pageY;
            style.transform = `translate3d(${x}px, ${y}px, 0`;
          }

          return (
            <Tooltip
              {...{
                arrowProps,
                closeOnOutOfBoundaries,
                outOfBoundaries,
                placement,
                scheduleUpdate,
                style,
                tooltip,
                trigger
              }}
              clearScheduled={this.clearScheduled}
              hideTooltip={this.hideTooltip}
              innerRef={ref}
            />
          );
        }}
      </Popper>
    );

    return (
      <Manager>
        <Reference innerRef={getTriggerRef}>
          {({ref}) =>
            children({getTriggerProps: this.getTriggerProps, triggerRef: ref})
          }
        </Reference>
        {this.getState() &&
          (usePortal ? createPortal(popper, portalContainer) : popper)}
      </Manager>
    );
  }

  private isControlled() {
    return this.props.tooltipShown !== undefined;
  }

  private getState() {
    return this.isControlled()
      ? this.props.tooltipShown
      : this.state.tooltipShown;
  }

  private setTooltipState = (state: TooltipTriggerState) => {
    const cb = () => this.props.onVisibilityChange(state.tooltipShown);
    this.isControlled() ? cb() : this.setState(state, cb);
  };

  private clearScheduled = () => {
    clearTimeout(this.hideTimeout);
    clearTimeout(this.showTimeout);
  };

  private showTooltip = ({pageX, pageY}: {pageX: number; pageY: number}) => {
    this.clearScheduled();
    let state: TooltipTriggerState = {
      tooltipShown: true
    };
    if (this.props.followCursor) {
      state = {
        ...state,
        pageX,
        pageY
      };
    }
    this.showTimeout = window.setTimeout(
      () => this.setTooltipState(state),
      this.props.delayShow
    );
  };

  private hideTooltip = () => {
    this.clearScheduled();
    this.hideTimeout = window.setTimeout(
      () => this.setTooltipState({tooltipShown: false}),
      this.props.delayHide
    );
  };

  private toggleTooltip = ({pageX, pageY}: {pageX: number; pageY: number}) => {
    const action = this.getState() ? 'hideTooltip' : 'showTooltip';
    this[action]({pageX, pageY});
  };

  private clickToggle: React.MouseEventHandler = (event: React.MouseEvent) => {
    event.preventDefault();
    const {pageX, pageY} = event;
    const action = this.props.followCursor ? 'showTooltip' : 'toggleTooltip';
    this[action]({pageX, pageY});
  };

  private contextMenuToggle: React.MouseEventHandler = event => {
    event.preventDefault();
    const {pageX, pageY} = event;
    const action = this.props.followCursor ? 'showTooltip' : 'toggleTooltip';
    this[action]({pageX, pageY});
  };

  private isTriggeredBy(event: TriggerTypes) {
    const {trigger} = this.props;
    return (
      trigger === event || (Array.isArray(trigger) && trigger.includes(event))
    );
  }

  private getTriggerProps = (props: GetTriggerPropsArg = {}) => {
    return {
      ...props,
      ...(this.isTriggeredBy('click') && {
        onClick: callAll(this.clickToggle, props.onClick),
        onTouchEnd: callAll(this.clickToggle, props.onTouchEnd)
      }),
      ...(this.isTriggeredBy('right-click') && {
        onContextMenu: callAll(this.contextMenuToggle, props.onContextMenu)
      }),
      ...(this.isTriggeredBy('hover') && {
        onMouseEnter: callAll(this.showTooltip, props.onMouseEnter),
        onMouseLeave: callAll(this.hideTooltip, props.onMouseLeave),
        ...(this.props.followCursor && {
          onMouseMove: callAll(this.showTooltip, props.onMouseMove)
        })
      }),
      ...(this.isTriggeredBy('focus') && {
        onFocus: callAll(this.showTooltip, props.onFocus),
        onBlur: callAll(this.hideTooltip, props.onBlur)
      })
    };
  };
}

export default TooltipTrigger;
