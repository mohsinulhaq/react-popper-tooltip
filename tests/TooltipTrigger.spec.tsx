import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import TooltipTrigger from '../src';

interface BasicTooltipTriggerProps {
  tooltip: React.ReactNode;
  children: React.ReactNode;
  hideArrow?: boolean;
  [key: string]: any;
}

const BasicTooltipTrigger = ({
  tooltip,
  children,
  hideArrow,
  ...props
}: BasicTooltipTriggerProps) => (
  <TooltipTrigger
    {...props}
    tooltip={({
      arrowRef,
      tooltipRef,
      getArrowProps,
      getTooltipProps,
      placement
    }) => (
      <div
        {...getTooltipProps({ref: tooltipRef, className: 'tooltip-container'})}
      >
        {!hideArrow && (
          <div
            {...getArrowProps({
              className: 'tooltip-arrow',
              'data-placement': placement,
              ref: arrowRef
            })}
          />
        )}
        {tooltip}
      </div>
    )}
  >
    {({getTriggerProps, triggerRef}) => (
      <span {...getTriggerProps({ref: triggerRef})}>{children}</span>
    )}
  </TooltipTrigger>
);

const Tooltip = 'Tooltip';
const Trigger = 'Trigger';
const Trigger2 = 'Trigger2';
const Outside = 'Outside';

window.MutationObserver = class {
  public disconnect() {}
  public observe() {}
};

jest.useFakeTimers();

it('matches snapshot', () => {
  render(
    <BasicTooltipTrigger defaultTooltipShown tooltip={Tooltip}>
      {Trigger}
    </BasicTooltipTrigger>
  );
  expect(document.body).toMatchSnapshot();
});

describe('hover trigger', () => {
  let container: HTMLElement;
  let queryByText: any;

  beforeEach(() => {
    ({container, queryByText} = render(
      <BasicTooltipTrigger tooltip={Tooltip}>{Trigger}</BasicTooltipTrigger>
    ));
    fireEvent.mouseEnter(container.firstChild as HTMLElement);
    jest.runAllTimers();
  });

  it('opens tooltip on mouseEnter', () => {
    expect(queryByText(Tooltip)).toBeTruthy();
  });

  it('closes tooltip on mouseLeave', () => {
    fireEvent.mouseLeave(container.firstChild as HTMLElement);
    jest.runAllTimers();
    expect(queryByText(Tooltip)).toBeFalsy();
  });

  it('retains tooltip after hovering into it', () => {
    fireEvent.mouseLeave(container.firstChild as HTMLElement);
    fireEvent.mouseEnter(queryByText(Tooltip));
    jest.runAllTimers();
    expect(queryByText(Tooltip)).toBeTruthy();
  });
});

describe('click trigger', () => {
  let container: HTMLElement;
  let queryByText: any;

  beforeEach(() => {
    ({container, queryByText} = render(
      <BasicTooltipTrigger trigger="click" tooltip={Tooltip}>
        {Trigger}
      </BasicTooltipTrigger>
    ));
    fireEvent.click(container.firstChild as HTMLElement);
    jest.runAllTimers();
  });

  it('opens tooltip on click', () => {
    expect(queryByText(Tooltip)).toBeTruthy();
  });

  it('closes tooltip on click', () => {
    fireEvent.click(container.firstChild as HTMLElement);
    jest.runAllTimers();
    expect(queryByText(Tooltip)).toBeFalsy();
  });
});

describe('right-click trigger', () => {
  let container: HTMLElement;
  let queryByText: any;

  beforeEach(() => {
    ({container, queryByText} = render(
      <BasicTooltipTrigger trigger="right-click" tooltip={Tooltip}>
        {Trigger}
      </BasicTooltipTrigger>
    ));
    fireEvent.contextMenu(container.firstChild as HTMLElement);
    jest.runAllTimers();
  });

  it('opens tooltip on rightClick', () => {
    expect(queryByText(Tooltip)).toBeTruthy();
  });

  it('closes tooltip on rightClick', () => {
    fireEvent.contextMenu(container.firstChild as HTMLElement);
    jest.runAllTimers();
    expect(queryByText(Tooltip)).toBeFalsy();
  });

  it('closes tooltip on click', () => {
    fireEvent.click(container.firstChild as HTMLElement);
    jest.runAllTimers();
    expect(queryByText(Tooltip)).toBeFalsy();
  });
});

describe('follow cursor', () => {
  let container: HTMLElement;
  let queryByText: any;

  beforeEach(() => {
    ({container, queryByText} = render(
      <BasicTooltipTrigger followCursor tooltip={Tooltip}>
        {Trigger}
      </BasicTooltipTrigger>
    ));
    fireEvent.mouseMove(container.firstChild as HTMLElement);
    jest.runAllTimers();
  });

  it('opens tooltip on mouseMove', () => {
    expect(queryByText(Tooltip)).toBeTruthy();
  });

  it('closes tooltip on mouseLeave', () => {
    fireEvent.mouseLeave(container.firstChild as HTMLElement);
    jest.runAllTimers();
    expect(queryByText(Tooltip)).toBeFalsy();
  });
});

it('closes on outside click', () => {
  const {container, queryByText} = render(
    <>
      <BasicTooltipTrigger trigger="click" tooltip={Tooltip}>
        {Trigger}
      </BasicTooltipTrigger>
      <div>{Outside}</div>
    </>
  );

  fireEvent.click(container.firstChild as HTMLElement);
  jest.runAllTimers();

  fireEvent.click(queryByText(Outside)!);
  jest.runAllTimers();

  expect(queryByText(Tooltip)).toBeFalsy();
});

describe('nested tooltips', () => {
  describe('correct opening', () => {
    it('works for click trigger', () => {
      const {queryByText} = render(
        <BasicTooltipTrigger
          trigger="click"
          tooltip={
            <BasicTooltipTrigger trigger="click" tooltip={Tooltip}>
              {Trigger2}
            </BasicTooltipTrigger>
          }
        >
          {Trigger}
        </BasicTooltipTrigger>
      );
      fireEvent.click(queryByText(Trigger)!);
      jest.runAllTimers();
      fireEvent.click(queryByText(Trigger2)!);
      jest.runAllTimers();
      expect(queryByText(Tooltip)).toBeTruthy();
    });

    it('works for hover trigger', () => {
      const {queryByText} = render(
        <BasicTooltipTrigger
          tooltip={
            <BasicTooltipTrigger tooltip={Tooltip}>
              {Trigger2}
            </BasicTooltipTrigger>
          }
        >
          {Trigger}
        </BasicTooltipTrigger>
      );
      fireEvent.mouseEnter(queryByText(Trigger)!);
      jest.runAllTimers();
      fireEvent.mouseEnter(queryByText(Trigger2)!);
      jest.runAllTimers();
      expect(queryByText(Tooltip)).toBeTruthy();
    });
  });

  describe('partial closure of nested tooltips', () => {
    it('works for click trigger', () => {
      const {queryByText} = render(
        <BasicTooltipTrigger
          trigger="click"
          tooltip={
            <BasicTooltipTrigger trigger="click" tooltip={Tooltip}>
              {Trigger2}
            </BasicTooltipTrigger>
          }
        >
          {Trigger}
        </BasicTooltipTrigger>
      );
      fireEvent.click(queryByText(Trigger)!);
      jest.runAllTimers();
      fireEvent.click(queryByText(Trigger2)!);
      jest.runAllTimers();
      fireEvent.click(queryByText(Trigger2)!);
      jest.runAllTimers();
      expect(queryByText(Tooltip)).toBeFalsy();
      expect(queryByText(Trigger2)).toBeTruthy();
    });

    it('works for hover trigger', () => {
      const {queryByText} = render(
        <BasicTooltipTrigger
          tooltip={
            <BasicTooltipTrigger tooltip={Tooltip}>
              {Trigger2}
            </BasicTooltipTrigger>
          }
        >
          {Trigger}
        </BasicTooltipTrigger>
      );
      fireEvent.mouseEnter(queryByText(Trigger)!);
      jest.runAllTimers();
      fireEvent.mouseEnter(queryByText(Trigger2)!);
      jest.runAllTimers();
      fireEvent.mouseLeave(queryByText(Trigger2)!);
      fireEvent.mouseEnter(queryByText(Trigger)!);
      jest.runAllTimers();
      expect(queryByText(Tooltip)).toBeFalsy();
      expect(queryByText(Trigger2)).toBeTruthy();
    });
  });

  describe('complete closure of nested tooltips', () => {
    it('works for click trigger', () => {
      const {queryByText} = render(
        <>
          <BasicTooltipTrigger
            trigger="click"
            tooltip={
              <BasicTooltipTrigger trigger="click" tooltip={Tooltip}>
                {Trigger2}
              </BasicTooltipTrigger>
            }
          >
            {Trigger}
          </BasicTooltipTrigger>
          <div>{Outside}</div>
        </>
      );
      fireEvent.click(queryByText(Trigger)!);
      jest.runAllTimers();
      fireEvent.click(queryByText(Trigger2)!);
      jest.runAllTimers();
      fireEvent.click(queryByText(Outside)!);
      jest.runAllTimers();
      expect(queryByText(Tooltip)).toBeFalsy();
      expect(queryByText(Trigger2)).toBeFalsy();
    });

    it('works for hover trigger', () => {
      const {queryByText} = render(
        <>
          <BasicTooltipTrigger
            tooltip={
              <BasicTooltipTrigger tooltip={Tooltip}>
                {Trigger2}
              </BasicTooltipTrigger>
            }
          >
            {Trigger}
          </BasicTooltipTrigger>
        </>
      );
      fireEvent.mouseEnter(queryByText(Trigger)!);
      jest.runAllTimers();
      fireEvent.mouseEnter(queryByText(Trigger2)!);
      jest.runAllTimers();
      fireEvent.mouseOut(queryByText(Trigger)!);
      jest.runAllTimers();
      expect(queryByText(Tooltip)).toBeFalsy();
      expect(queryByText(Trigger2)).toBeFalsy();
    });
  });
});
