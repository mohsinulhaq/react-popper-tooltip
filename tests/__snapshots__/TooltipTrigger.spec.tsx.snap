// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<body>
  <div>
    <span>
      Trigger
    </span>
  </div>
  <div
    class="tooltip-container"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="tooltip-arrow"
      data-placement="right"
    />
    Tooltip
  </div>
</body>
`;
